x/emissions/module/rewards/topic_rewards.go

```diff
@@ -53,7 +53,7 @@ func SafeApplyFuncOnAllRewardReadyTopics(
 	fn func(ctx context.Context, topic *types.Topic) error,
 	topicPageLimit uint64,
 	maxTopicPages uint64,
-) error {
+	filterRewardNonces bool) error {
```

Comment:
Instead of this and the other bool you added elsewhere probs have to "flip the relationship" between
`SafeApplyFuncOnAllRewardReadyTopics`
and
`ApplyFuncOnAllChurnReadyTopics`

the latter calls `GetTopicWeights` which calls the former. Meanwhile, reward ready topics are a subset of churnable topics

Reply:
However we first need to get all topic weights anyway, so we can get the topN (because that needs to be calculated on the fly).

The alternative would be to remove those filters, but because the filters are coupled with running a particular function on that data, so then to enable/disable filters.

Suggestion: We can decouple those `*ApplyFuncOnAll*ReadyTopics` functions into parts: return the data first, then do any particular filtering or action on each case (filtering: apply on rewardReady on nonces, no filter on churnReady; action: set previous weight on reWradReady and no action on churnReady).
This makes the separation explicit and easier to read.

Reply:
What you're suggesting ends up looking more complex than you think, because you need to paginate through all entities wherever you apply this logic now. Meanwhile, before, when it was all part of `*ApplyFuncOnAll*ReadyTopics`, you have all that pagination functionality abstracted away.

[Cosmos docs](https://docs.cosmos.network/v0.50/build/packages/collections#example-3) recommends it this way anyway, albeit with fewer safeguards than we have.